[@property] Fallback that does not match syntax should make var() invalid

anttijk opened the pull request:
#### 6dabb8f934a3f64c639eb12d8c24ddea1c16101e
<pre>
[@property] Fallback that does not match syntax should make var() invalid
<a href="https://bugs.webkit.org/show_bug.cgi?id=249981">https://bugs.webkit.org/show_bug.cgi?id=249981</a>
rdar://103799338

Reviewed by Sam Weinig.

The fallback value must match the syntax definition of the custom property being referenced,
otherwise the declaration is invalid at computed-value time.

<a href="https://drafts.css-houdini.org/css-properties-values-api/#fallbacks-in-var-references">https://drafts.css-houdini.org/css-properties-values-api/#fallbacks-in-var-references</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/var-reference-registered-properties-expected.txt:
* Source/WebCore/css/CSSVariableReferenceValue.cpp:
(WebCore::CSSVariableReferenceValue::resolveVariableFallback const):

Check the syntax of the fallback.

(WebCore::CSSVariableReferenceValue::resolveVariableReference const):
(WebCore::CSSVariableReferenceValue::resolveTokenRange const):
(WebCore::CSSVariableReferenceValue::resolveVariableReferences const):
(WebCore::CSSVariableReferenceValue::resolveVariableFallback): Deleted.
(WebCore::CSSVariableReferenceValue::resolveVariableReference): Deleted.
(WebCore::CSSVariableReferenceValue::resolveTokenRange): Deleted.

Make these non-static (so parser context is available) and did some return value cleanup.

* Source/WebCore/css/CSSVariableReferenceValue.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::isValidCustomPropertyValueForSyntax):
* Source/WebCore/css/parser/CSSPropertyParser.h:

Canonical link: <a href="https://commits.webkit.org/258373@main">https://commits.webkit.org/258373@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9edf1da8fca893a7bed625808dd93e83b8e33fa5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101700 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10850 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34783 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111034 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/171238 "Found 1 new test failure: js/dom/Promise-reject-large-string.html (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105680 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11806 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1761 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94108 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108796 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107480 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9022 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/92291 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/36649 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90915 "Found 1 new API test failure: TestWebKitAPI.WebKit.QuotaDelegateReload (failure)") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/23718 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78586 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4448 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/25210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4523 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1647 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10611 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/44697 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5738 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6277 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/css/CSSVariableReferenceValue.cpp:
```diff
@@ -139,16 +158,19 @@ bool CSSVariableReferenceValue::resolveTokenRange(CSSParserTokenRange range, Vec
         return nullptr;
 
-    return CSSVariableData::create(resolvedTokens);
+    return CSSVariableData::create(*resolvedTokens);
```
Seems unfortunate we have to copy here. Can we add a CSSVariableData::create() overload that takes an r-value and move the resolved tokens in? Or does the processing done in CSSVariableData's constructor make this a meaning less optimization?

anttijk replied:
Maybe something to do separately. I did notice it and wondered about the same things you did.

anttijk committed:
[@property] Fallback that does not match syntax should make var() invalid
https://bugs.webkit.org/show_bug.cgi?id=249981
rdar://103799338

Reviewed by Sam Weinig.

The fallback value must match the syntax definition of the custom property being referenced,
otherwise the declaration is invalid at computed-value time.

https://drafts.css-houdini.org/css-properties-values-api/#fallbacks-in-var-references

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/var-reference-registered-properties-expected.txt:
* Source/WebCore/css/CSSVariableReferenceValue.cpp:
(WebCore::CSSVariableReferenceValue::resolveVariableFallback const):

Check the syntax of the fallback.

(WebCore::CSSVariableReferenceValue::resolveVariableReference const):
(WebCore::CSSVariableReferenceValue::resolveTokenRange const):
(WebCore::CSSVariableReferenceValue::resolveVariableReferences const):
(WebCore::CSSVariableReferenceValue::resolveVariableFallback): Deleted.
(WebCore::CSSVariableReferenceValue::resolveVariableReference): Deleted.
(WebCore::CSSVariableReferenceValue::resolveTokenRange): Deleted.

Make these non-static (so parser context is available) and did some return value cleanup.

* Source/WebCore/css/CSSVariableReferenceValue.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::isValidCustomPropertyValueForSyntax):
* Source/WebCore/css/parser/CSSPropertyParser.h:

Canonical link: https://commits.webkit.org/258373@main
